includes/class-amp-story-post-type.php

```diff
@@ -1490,6 +1522,232 @@ public static function insert_attachment( $object, $cropped ) {
 		return $attachment_id;
 	}
 
+	/**
+	 * Ajax handler to export the story ZIP archive.
+	 *
+	 * This method returns an error as JSON and the binary data on success.
+	 */
+	public static function handle_export() {
+		check_ajax_referer( 'amp-story-export', 'nonce' );
+
+		// The user must have the correct permissions.
+		if ( ! current_user_can( 'edit_files' ) ) {
+			wp_send_json_error(
+				array(
+					'errorMessage' => esc_html__( 'You do not have the required permissions to export AMP stories.', 'amp' ),
+				),
+				403
+			);
+		}
+
+		// We need the ZipArchive class to make this work.
+		if ( ! class_exists( 'ZipArchive', false ) ) {
+			wp_send_json_error(
+				array(
+					'errorMessage' => esc_html__( 'The ZipArchive class is required to export AMP stories.', 'amp' ),
+				),
+				403
+			);
+		}
+
+		// Bail if the user has not saved the story yet.
+		if ( 'auto-draft' === get_post_status( wp_unslash( $_POST['post_ID'] ) ) ) {
+			wp_send_json_error(
+				array(
+					'errorMessage' => esc_html__( 'Save the AMP story before exporting.', 'amp' ),
+				),
+				403
+			);
+		}
+
+		// Generate and export the archive.
+		$export = self::generate_export( wp_unslash( $_POST['post_ID'] ) );
+
+		// Export failed.
+		if ( is_wp_error( $export ) ) {
+			wp_send_json_error(
+				array(
+					'errorMessage' => $export->get_error_message(),
+				),
+				403
+			);
+		}
+
+		// Failed to export for an unknown reason not related to generating the archive.
+		wp_send_json_error(
+			array(
+				'errorMessage' => __( 'Could not generate the AMP story archive.', 'amp' ),
+			),
+			403
+		);
+	}
+
+	/**
+	 * Generates a Zip archive from the AMP Story.
+	 *
+	 * @param int $post_id The post ID of the AMP Story.
+	 * @return WP_Error
+	 */
+	private static function generate_export( $post_id ) {
+		$post = get_post( $post_id );
+
+		if ( ! $post ) {
+			return new WP_Error( 'amp_story_export_invalid_post', esc_html__( 'The AMP Story does not exist.', 'amp' ) );
+		}
+
+		$slug = sanitize_title( $post->post_title, $post->ID );
+		$file = sprintf( '%s.zip', $slug );
+
+		$zip = new ZipArchive();
+		$res = $zip->open( $file, ZipArchive::CREATE | ZipArchive::OVERWRITE );
+
+		// Generate ZIP and send headers.
+		if ( true === $res ) {
+
+			// Create the zip directory.
+			$zip->addEmptyDir( $slug );
+
+			// Add README.txt file.
+			$zip->addFromString( $slug . '/README.txt', 'temporary content...' );
+
+			// Passed to `get_preview_post_link()` for nonce access and to sanitize the output.
+			$query_args = array(
+				'story_export' => $slug,
+				'_wpnonce'     => wp_create_nonce( 'amp-story-export-' . $post->ID ),
+			);
+
+			// Passed to `wp_remote_get()`.
+			$args = array(
+				'timeout'   => 30,
+				'sslverify' => false,
+			);
+
+			// Get the preview URL.
+			$response = wp_remote_get( get_preview_post_link( $post, $query_args ), $args );
```

Comment:
Just thinking that the cookies in the current request may need to be passed along here, so that draft stories and private stories can be exported. You can refer to the validation manager for example code here:

https://github.com/ampproject/amp-wp/blob/9342ddab702c06ad496cd36ff1f21f30ffc12014/includes/validation/class-amp-validation-manager.php#L1833-L1844

Reply:
Additionally, for doing nonce checking for unauthenticated users, refer also to:

https://github.com/ampproject/amp-wp/blob/9342ddab702c06ad496cd36ff1f21f30ffc12014/includes/validation/class-amp-validation-manager.php#L1633-L1634

https://github.com/ampproject/amp-wp/blob/9342ddab702c06ad496cd36ff1f21f30ffc12014/includes/validation/class-amp-validation-manager.php#L1822-L1826

Reply:
Thank you for the direction here. I was able to rework this.